### PR TITLE
Add vector-Jacobian-products for a subset of nvFuser-supported prims; add backward support for nvprims

### PR DESCRIPTION
**Pull request by IvanYashchuk**
NOTE: Currently, this PR is blocked until https://github.com/pytorch/pytorch/issues/85696 is resolved.

`vjp_implementations["prim_name"]` is a callable that computes vector-Jacobian products for given cotangents, forward results, and forward inputs. And this function is used in the `torch.autograd.Function.backward` method.

Here's an example trace for the `torch.sigmoid` function:
```py
import torch
from torch._prims.context import NvfuserPrimsMode, TorchRefsMode
from torch.fx.experimental.proxy_tensor import make_fx

a = torch.randn(3, 3, device='cuda', requires_grad=True)
g = torch.ones_like(a)

def func(a):
    c = torch.sigmoid(a)
    return torch.autograd.grad(c, a, g)

# torch._refs.sigmoid trace
with NvfuserPrimsMode(), TorchRefsMode():
    make_fx(func)(a).graph.print_tabular()
```
```
opcode         name                 target                     args                        kwargs
-------------  -------------------  -------------------------  --------------------------  --------
placeholder    a_1                  a_1                        ()                          {}
call_function  neg                  nvprims.neg.default        (a_1,)                      {}
call_function  detach               aten.detach.default        (neg,)                      {}
call_function  exp                  nvprims.exp.default        (neg,)                      {}
call_function  detach_1             aten.detach.default        (exp,)                      {}
call_function  add                  nvprims.add.default        (1.0, exp)                  {}
call_function  detach_2             aten.detach.default        (add,)                      {}
call_function  div                  nvprims.div.default        (1.0, add)                  {}
call_function  detach_3             aten.detach.default        (div,)                      {}
get_attr       _tensor_constant0    _tensor_constant0          ()                          {}
call_function  is_same_size         aten.is_same_size.default  (div, _tensor_constant0)    {}
call_function  detach_4             aten.detach.default        (detach_3,)                 {}
call_function  detach_5             aten.detach.default        (detach_3,)                 {}
get_attr       _tensor_constant0_1  _tensor_constant0          ()                          {}
call_function  div_1                nvprims.div.default        (_tensor_constant0_1, add)  {}
get_attr       _tensor_constant0_2  _tensor_constant0          ()                          {}
call_function  neg_1                nvprims.neg.default        (_tensor_constant0_2,)      {}
call_function  mul                  nvprims.mul.default        (neg_1, 1.0)                {}
call_function  pow_1                nvprims.pow.default        (add, -2)                   {}
call_function  mul_1                nvprims.mul.default        (mul, pow_1)                {}
call_function  detach_6             aten.detach.default        (detach_2,)                 {}
call_function  detach_7             aten.detach.default        (detach_2,)                 {}
call_function  view_of              nvprims.view_of.default    (mul_1,)                    {}
call_function  view_of_1            nvprims.view_of.default    (mul_1,)                    {}
call_function  detach_8             aten.detach.default        (detach_1,)                 {}
call_function  detach_9             aten.detach.default        (detach_1,)                 {}
call_function  mul_2                nvprims.mul.default        (view_of_1, detach_9)       {}
call_function  detach_10            aten.detach.default        (detach,)                   {}
call_function  detach_11            aten.detach.default        (detach,)                   {}
call_function  neg_2                nvprims.neg.default        (mul_2,)                    {}
output         output               output                     ((neg_2,),)                 {}
```

`aten.detach` and `aten.is_same_size` appear in the trace that are called somewhere in the autograd engine code but other than that the trace consists of `nvprims` calls.

Thanks to @rdspring1 for implementing several VJP functions!

cc @kevinstephano @jjsjann123 @ezyang @mruberry @ngimel @Lezcano @fdrocha @peterbell10

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
@rdspring1, can't request you as a reviewer for some reason, but could you please take a look at this PR?

File: torch/_prims/vjp.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+# add dimensions in canonicalized, sorted order
+# so that the tensor has the appropriate rank.
+def _unsqueeze_dims(a, dims, rank):
+    dims = utils.canonicalize_dims(rank, dims)
+    for dim in sorted(dims):
+        a = torch._refs.unsqueeze(a, dim)
+    return a
+
+
+def _dim_size(a, dims):
```
Comment (mruberry) [inline torch/_prims/vjp.py:22]: Add comments for these functions

File: torch/utils/_mode_utils.py [hunk 2]
```diff
@@ -15,3 +15,11 @@ def no_dispatch():
         yield
     finally:
         del guard
+
+@contextmanager
```
Comment (mruberry) [inline torch/utils/_mode_utils.py:19]: Add a comment for this context manager

File: torch/_prims/vjp.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
+# Module for defining vector-Jacobian product (VJP) functions for PrimTorch
```
Comment (mruberry) [inline torch/_prims/vjp.py:1]: Should we define the vjps alongside the prims themselves? Seems like a fundamental transform

File: test/test_nvfuser_prims.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        op.op = partial(make_traced(op.op), executor="aten")
+
+        self._check_helper(device, dtype, op, op.get_op())
+
+    @onlyCUDA
+    @skipCUDAIfRocm
+    @ops(ref_nvprims_ops, dtypes=OpDTypes.supported, allowed_dtypes=(torch.float64,))
+    def test_nvprims_grad_trace_backward(self, device, dtype, op):
+        # Check if result is boolean
+        if op.torch_opinfo_name in ["eq", "ne", "gt", "ge", "lt", "le", "isfinite"]:
```
Comment (mruberry) [inline test/test_nvfuser_prims.py:145]: Instead of checking names use skips on the OpInfos, use the OpInfo metadata for whether the op supports autograd or not, or directly test whether the output is not a differentiable datatype

File: torch/_prims/vjp.py (hunk 3 above)
Comment (IvanYashchuk) [reply torch/_prims/vjp.py:1]: Maybe we should. I did that initially in https://github.com/pytorch/pytorch/pull/80448 by adding `impl_vjp=` alongside prims.

File: torch/_prims/vjp.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+import torch
+import torch._prims_common as utils
+
+prims = torch.ops.prims
+
+
+# add dimensions in canonicalized, sorted order
+# so that the tensor has the appropriate rank.
+def _unsqueeze_dims(a, dims, rank):
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply torch/_prims/vjp.py]: No reason, I suppose it was taken directly from derivatives.yaml. `_prims.expand_dims` also works.
